torch/distributed/_mesh_layout.py

```diff
@@ -0,0 +1,337 @@
+            sub_stride = numel_so_far
+            numel_so_far *= sub_size
+            sub_res_sizes, sub_res_strides = [], []
+
+            # when self is multi-dimensional sublayout, aka, self = (a,b,...,c):(x,y,...,z), layout = s:d,
+            # for integral s and d means that we want:
+            # (1) “remove” the first d elements from self. (This will increase the stride.)
+            # (2) “keep” the first s of those strided elements. (This does not affect the stride.)
+            # For example, if self = (6,2):(2,1), layout = (3:2)
+            # Step 1: remove the first 2 elements from self with strid increase, i.e., (6,2):(2,1) -> (3,2):(4,1)
+            # Step 2: keep the first 3 of those strided elements, i.e., (3,2):(4,1) -> (3,1):(4,1)
+            for curr_size, curr_stride in zip(self.sizes[:-1], self.strides[:-1]):
+                assert curr_size % sub_stride == 0 or sub_stride % curr_size == 0, (
+                    "Layouts do not meet stride divisibility condition"
+                )
+                new_size = min(max(1, curr_size // sub_stride), sub_size)
+                if new_size != 1:
+                    sub_res_sizes.append(new_size)
+                    sub_res_strides.append(sub_stride * curr_stride)
+                assert sub_size % new_size == 0, (
+                    "Layouts do not meet shape divisibility condition"
+                )
+                sub_size = sub_size // new_size
+                sub_stride = self.ceil_div(sub_stride, curr_size)
+
+            # When self is integral and has single-size sublayout, aka, self = a:b, layout = s:d,
+            # the result is rather trivial: self o layout = a:b o s:d = s:(b*d).
+            # For example, if self = (6:2), layout = (3:2), the result is (3:(2*2)) = (3:4).
+            if sub_size != 1 or len(sub_res_sizes) == 0:
+                sub_res_sizes.append(sub_size)
+                sub_res_strides.append(sub_stride * self.strides[-1])
+
+            sub_res = _Layout(tuple(sub_res_sizes), tuple(sub_res_strides))
+            res.append(sub_res)
+        return res
+
+    def complement(self, world_size: int) -> "_Layout":
+        """
+        Compute the "complement layout" relative to a given world_size.
+        A complement layout fills in the "missing" factor so that: self repeat a layout of complement(self, world_size)
+        will get a total cosize (product of sizes) = world_size. We use ⊗ to denote the repeat operation.
+
+        Example:
+          self = (4:1)   # size=4, stride=1
+          world_size = 8
+          Then:
+            cosize(self) = 4
+            Needed factor = 8 / 4 = 2
+            complement(self, 8) = (2:1)
+
+          Together they form:
+            (4:1) ⊗ (2:1) = (4,2):(2,1)
+          which has cosize = 4 * 2 = 8, as required.
+
+        In distributed terms, complement() is often used to derive the "other"
+        rank grouping when splitting processes into 2D meshes.
```

Comment:
This is one of the things I want to understand. (It would also be useful to have equivalent "In distributed" comments for coalesce and composition. Those are easier operations to understand but it's not immediately clear why they payoff. (BTW, after writing this I'm going to carefully study the next PR in this stack to see if I can work it out.)

Reply:
I understood it, and tweeted about it here https://x.com/ezyang/status/1962364978393981433

Reply:
Do you want me to include your notes inside comment so that we have a reference to make people understand it easier down the road.